subxt/tests/integration/utils/node_proc.rs

```diff
@@ -185,25 +150,30 @@ impl TestNodeProcessBuilder {
     }
 }
 
-/// Returns the next set of 3 open ports.
-///
-/// Returns None if there are not 3 open ports available.
-fn next_open_port() -> Option<(u16, u16, u16)> {
-    // Ask the kernel to allocate a port.
-    let next_port = || {
-        match TcpListener::bind(("127.0.0.1", 0)) {
-            Ok(listener) => {
-                if let Ok(address) = listener.local_addr() {
-                    Some(address.port())
-                } else {
-                    None
-                }
-            }
-            Err(_) => None,
-        }
-    };
-
-    // The ports allocated should be different, unless in
-    // the unlikely case that the system has less than 3 available ports.
-    Some((next_port()?, next_port()?, next_port()?))
+// Consume a stderr reader from a spawned substrate command and
+// locate the port number that is logged out to it.
+fn find_substrate_port_from_output(r: impl Read + Send + 'static) -> u16 {
+    BufReader::new(r)
+        .lines()
+        .find_map(|line| {
+            let line = line
+                .expect("failed to obtain next line from stdout for port discovery");
+
+            // does the line contain our port (we expect this specific output from substrate).
+            let line_end = match line.rsplit_once("Listening for new connections on 127.0.0.1:") {
```

Comment:
this will not work once we update to jsonrpsee in substrate but a later question I guess.

Reply:
DQ: Why wont this work with the update? Just out of curiousity.

Reply:
Aah, different log output for the port I guess?

(I wonder what the log output will be? it should be easy enough to look for both when the time comes, to continue to support older nodes :))

(I do wish there was a better way to find the port that a thing is running on; maybe I should look into this a bit more!)

Reply:
Never tried it, but maybe you could use something like https://docs.rs/netstat/latest/netstat/ to find the open ports of the spawned process, then might be easy enough to find the websockets port?

Reply:
Ooh thanks for the pointer; I might have a look into this!

Reply:
> Aah, different log output for the port I guess?

Yeah, but we could change to have this log in jsonrpsee it's quite useful.